demo.py

```diff
@@ -0,0 +1,146 @@
+class Resource(models.Model):
+    client = ForeignKey(Client)
+    owned_by = ForgeignKey(Resource, null=True)
+    name = CharField(blank=False, max_length=128)
+    configuration = JSONField(null=False)
+    types = ResourceTypeRegistry()
+    owned_resources = ResourceManger()
+
+    @property
+    def configuration_schema(self):
+        """ Define the options used to configure this resource. """
+        return {}  # No configurable options by default
+
+    @property
+    def state(self):
+        """
+        The current state of this resource.
+        The return value is always a subclass of ResourceState, and typically will be a subclass of:
+          NeedsRefreshState, ApplyingConfigurationState, NeedsInputState, ReadyState, ErrorState, or DestroyingState
+        Each state supports common methods like get_status_message(), get_status_color(), allow_configuration_changes()
+        """
+        return self._state
+
+    def build_form():
+        """ Convert self.configuration_schema into a Django form """
+        pass
+
+
+class VmProvider(Resource):
+    """ A provider that can provision virtual machines """
+    class Meta:
+        proxy = True
```

Comment:
@bradenmacdonald Would adding new resource types be such a frequent occurrence that it needs to be optimised to that extent? Plus, that removes the django migrations, but the need to maintain and migrate that data isn't going to disappear - it means when the "model" of a configuration field is going to evolve, we'll have to do the work of django migrations ourselves.

For detecting configuration changes from children, that seem looks like something the dirty field pattern would solve well, no? There is even a lib for it: https://github.com/romgar/django-dirtyfields

For "seeing" the configuration of children, are you referring to the fact that queries on the Resource manager would return Resource objects, not children objects? Could https://django-model-utils.readthedocs.org/en/latest/managers.html#inheritancemanager help with this?

In a nutshell, I'm wondering if it wouldn't be better to reuse existing well-supported patterns, rather than reimplementing something entirely new of our own baking. Things like Django migrations, Model fields, ModelForms & co work well already, and putting ourselves outside of that pattern will necessarily generate work to replace it, taking the place of other work, which could be more directly linked to our current goals?